File: codes/rust/46.全排列.rs
/*
 * @lc app=leetcode.cn id=46 lang=rust
 *
 * [46] 全排列
 */

// @lc code=start
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut state = Vec::new();
        let mut selected = vec![false; nums.len()];
        let choices = nums;
        Self::backtrack(state, &choices, &mut selected, &mut result);
        result
    }

    fn backtrack(
        mut state: Vec<i32>,
        choices: &Vec<i32>,
        selected: &mut Vec<bool>,
        result: &mut Vec<Vec<i32>>,
    ) {
        // 如果最终状态的长度等于可选择的状态长度则添加结果并返回
        if state.len() == choices.len() {
            result.push(state);
            return;
        }
        // 开始选择数值
        for i in 0..choices.len() {
            // 查看是否被选择过
            if selected[i] {
                continue;
            }
            selected[i] = true;
            state.push(choices[i]);
            Self::backtrack(state.clone(), choices, selected, result);
            selected[i] = false;
            state.pop();
        }
    }
}
// @lc code=end

File: codes/golang/46.全排列.go
/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 */

import "fmt"

// @lc code=start
func permute(nums []int) [][]int {
	var res [][]int
	var trace []int
	backTrace(nums, &trace, &res)
	return res
}

func backTrace(nums []int, trace *[]int, res *[][]int) {

	if len(*trace) == len(nums) {
		temp := make([]int, len(*trace))
		copy(temp, *trace)
		*res = append(*res, temp)
		return
	}
	for i := 0; i < len(nums); i++ {
		if IsExistItem(nums[i], *trace) {
			continue
		}
		*trace = append(*trace, nums[i])
		backTrace(nums, trace, res)
		*trace = (*trace)[:len(*trace)-1]
	}
}

func IsExistItem(value int, array []int) bool {
	for _, v := range array {
		if v == value {
			return true
		}
	}
	return false
}

// @lc code=end

